app/models/experimentation/notifications_experiment.rb

```diff
@@ -79,9 +71,36 @@ def record_notification_results
       end
     end
 
+    def evict_patients
+      treatment_group_memberships.status_enrolled
+        .joins(:appointment)
+        .where("appointments.status <> 'scheduled' or appointments.remind_on > expected_return_date")
+        .evict(reason: "appointment_moved")
+
+      treatment_group_memberships.status_enrolled
+        .joins(patient: :latest_scheduled_appointments)
+        .where("treatment_group_memberships.appointment_id <> appointments.id")
+        .evict(reason: "new_appointment_created_after_enrollment")
+
+      treatment_group_memberships.status_enrolled
+        .joins(treatment_group: :reminder_templates)
+        .where("messages -> reminder_templates.message ->> 'result' = 'failed'")
+        .evict(reason: "notification_failed")
+
+      cancel_evicted_notifications
+    end
+
+    def schedule_notifications(date)
+      memberships_to_notify(date)
+        .select("reminder_templates.id reminder_template_id")
+        .select("reminder_templates.message message, treatment_group_memberships.*")
+        .in_batches(of: BATCH_SIZE)
+        .each_record { |membership| schedule_notification(membership, date) }
```

Comment:
Should this happen in a tx?

Reply:
Since this is idempotent, we don't need to put this in a transaction. We can re-run this in case of a failure.